gobblin-data-management/src/main/java/gobblin/data/management/retention/BaseDatasetCleaner.java

```diff
@@ -0,0 +1,215 @@
+ */
+public class BaseDatasetCleaner implements Instrumentable, Closeable {
+
+  public static final String CONFIGURATION_KEY_PREFIX = "gobblin.retention.";
+  public static final String DATASET_PROFILE_CLASS_KEY = CONFIGURATION_KEY_PREFIX + "dataset.finder.class";
+
+  public static final String MAX_CONCURRENT_DATASETS_CLEANED = CONFIGURATION_KEY_PREFIX
+      + "max.concurrent.datasets.cleaned";
+  public static final String DATASET_CLEAN_HDFS_CALLS_PER_SECOND_LIMIT = CONFIGURATION_KEY_PREFIX
+      + "hdfs.calls.per.second.limit";
+
+  public static final String DEFAULT_MAX_CONCURRENT_DATASETS_CLEANED = "1000";
+
+  private static Logger LOG = LoggerFactory.getLogger(DatasetCleaner.class);
+
+  protected List<DatasetFinder<Dataset>> datasetFinders;
+  protected final FileSystem optionalRateControlledFs;
+  private final ListeningExecutorService service;
+  private final Closer closer;
+  private final boolean isMetricEnabled;
+  private MetricContext metricContext;
+  private Optional<Meter> datasetsCleanSuccessMeter = Optional.absent();
+  private Optional<Meter> datasetsCleanFailureMeter = Optional.absent();
+  private Optional<CountDownLatch> finishCleanSignal;
+
+  public BaseDatasetCleaner(FileSystem fs, Properties jobProps) throws IOException {
+
+    this.closer = Closer.create();
+    try {
+
+      if (jobProps.contains(DATASET_CLEAN_HDFS_CALLS_PER_SECOND_LIMIT)) {
+        optionalRateControlledFs =
+            this.closer.register(new RateControlledFileSystem(fs, Long.parseLong(jobProps
+                .getProperty(DATASET_CLEAN_HDFS_CALLS_PER_SECOND_LIMIT))));
+        ((RateControlledFileSystem) optionalRateControlledFs).startRateControl();
+      } else {
+        this.optionalRateControlledFs = fs;
+      }
+    } catch (ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+
+    ExecutorService executor =
+        ScalingThreadPoolExecutor.newScalingThreadPool(0, Integer.parseInt(jobProps.getProperty(
+            MAX_CONCURRENT_DATASETS_CLEANED, DEFAULT_MAX_CONCURRENT_DATASETS_CLEANED)), 100, ExecutorsUtils
+            .newThreadFactory(Optional.of(LOG), Optional.of("Dataset-cleaner-pool-%d")));
+    this.service = MoreExecutors.listeningDecorator(executor);
+
+    // TODO -- Remove the dependency on gobblin-core after new Gobblin Metrics does not depend on gobblin-core.
+    this.metricContext = this.closer.register(Instrumented.getMetricContext(new State(jobProps), DatasetCleaner.class));
+    this.isMetricEnabled = GobblinMetrics.isEnabled(jobProps);
+  }
+
+  /**
+   * Perform the cleanup of old / deprecated dataset versions.
+   *
+   * @throws IOException
+   */
+  public void clean() throws IOException {
+
+    for (DatasetFinder<Dataset> datasetFinder : this.datasetFinders) {
+
+      try {
+        List<Dataset> dataSets = datasetFinder.findDatasets();
+        finishCleanSignal = Optional.of(new CountDownLatch(dataSets.size()));
```

Comment:
The class variable finishCleanSignal is used to track the overall cleaning process, right? but it got recreated for each datasetFinder, that does not sound right